forecast_tools.R

```diff
@@ -50,6 +50,90 @@ plot_data = function(data, title, observed) {
+#' 
+#' Will only return values where there are matching comparison columns (currency, level, species)
+#' 
+#' @param observations dataframe Has the columns NewMoonNumber, currency, level, species, actual
+#' @param forecasts dataframe passes the forecast validity check. Must have matching values in
+#'                  the comparison columns
+#' @param error_metric chr either 'mse' for mean squared error or 'likelihood' for the likelihood of the observation
+#' assumring the estimate and PI's fit a normal distribution
+#' @param ci_value int The value of the forecast confidence interval to scale PI values for the likelihood metric
+#' @return data.frame Data.frame with the columns model, error, lead_time, level, species, currency
+#' 
+calculate_forecast_error = function(observations, forecasts, error_metric='MSE', ci_value=90){
+  #The tibble datatype output from dplyr causes issues here
+  observations = as.data.frame(observations)
+  forecasts = as.data.frame(forecasts)
+  
+  if(!forecast_is_valid(forecasts)) stop('Forecast datafarme not valid')
+  
+  valid_observation_columns=c('NewMoonNumber','currency','level','species','actual')
+  if(!all(valid_observation_columns %in% colnames(observations))) stop('observation data.frame does not have valid column names')
+  
+  #At least 1 matching value must be in each of these columns in the observations and forecasts
+  #TODO: Ensure matching rows in all 3 columns at once instead of just one at a time.
+  column_check=c()
+  for(column in c('currency','level','species')){
+    if(!any(unique(observations[,column]) %in% unique(forecasts[,column]))) column_check=c(column_check, column)
+  }
+  if(length(column_check)>0) stop(paste('Comparison columns do not match: ',column_check, collaps=' '))
+  
+  #Calculate error
+  if(error_metric == 'MSE'){
+    comparison = forecasts %>%
+      inner_join(observations, by=c('NewMoonNumber','currency','level','species')) %>%
+      group_by(date, model, NewMoonNumber, currency, level, species) %>%
+      summarise(error=(estimate-actual)^2) %>%
+      ungroup()
+  } else if(error_metric == 'Likelihood') {
+    stop('Likelihood not implimented yet')
+  } else {
+    stop(paste0('Error metric unknown: ',error_metric))
+  }
+  
+  #Summarize to mean error by lead time. Lead time is number of new moons ahead of when the forecast was made.
+  #This assumes a forecast was made with only the data available prior to the first NewMoonDate in the series.
+  #TODO: Make the lead time the actual days or weeks once more frequent forecasts are being made( see #37)
+  forecast_date_new_moon_number = comparison %>%
+    group_by(date) %>%
+    summarise(new_moon_of_forecast = min(NewMoonNumber)-1) %>%
+    ungroup()
+  
+  comparison = comparison %>%
+    left_join(forecast_date_new_moon_number, by='date') %>%
+    mutate(lead_time=NewMoonNumber - new_moon_of_forecast) %>%
+    select(-new_moon_of_forecast, -NewMoonNumber, -date)
+  
+  comparison = comparison %>%
```

Comment:
Any reason not to have one continuous pipe here? If it's more readable split up I think having a separately named intermediate variable would make the logic easier to follow.